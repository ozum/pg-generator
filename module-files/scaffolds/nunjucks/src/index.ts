console.warn("This is a pg-generator plugin. It is not intended for direct usage. Use it vie pg-generator."); // eslint-disable-line no-console
