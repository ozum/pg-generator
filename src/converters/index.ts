export * from "./mermaid-to-svg";
