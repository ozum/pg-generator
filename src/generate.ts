/* eslint-disable no-nested-ternary */
import type { Options } from "./types";
import { getDb, composeWith } from "./utils";

/**
 * Executes the default sub-generator `app` from a generator.
 *
 * @param generator is the name or path of the generator. If it is a local path, use `require.resolve` or provide an absolute path.
 * @param options are the options for the generator.
 *
 * @example
 * generate("generator-from-npm", options);
 * generate(require.resolve("./local-generator"), options);
 */
export async function generate(generator: string, options?: Options): Promise<void>;
/**
 * Executes a sub-generator from a generator.
 *
 * @param generator is the name or path of the generator. If it is a local path, use `require.resolve` or provide an absolute path.
 * @param subGenerator is the name of the generator.
 * @param options are the new options added on top of curent options.
 *
 * @example
 * generate("generator-from-npm", "sub-generator");
 * generate(require.resolve("./local-generator"), "sub-generator", options);
 */
export async function generate(generator: string, subGenerator?: string, options?: Options): Promise<void>;
export async function generate(generator: string, subOrOptions?: string | Options, maybeOptions?: Options): Promise<void> {
  const options = typeof maybeOptions === "object" ? maybeOptions : typeof subOrOptions === "object" ? subOrOptions : {};
  const generatorName = typeof subOrOptions === "string" ? subOrOptions : undefined;
  const db = await getDb(options);
  return composeWith(generator, generatorName, options, db, process.cwd());
}
