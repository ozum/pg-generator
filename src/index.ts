export { PgGenerator } from "./pg-generator";
export { generate } from "./generate";
export { scaffold } from "./scaffold";

// Generator Helpers
export * as filterFunctions from "./utils/filter-functions";
export * as converters from "./converters";

// Types
export type { Context, GeneratorOptions, ClientOptions, Options } from "./types/index";
