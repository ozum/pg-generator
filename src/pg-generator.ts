/* eslint-disable class-methods-use-this */
import { extname, join, dirname, resolve } from "path";
import { promises as fs } from "fs";
import { chalk } from "meow-helper";
import merge from "lodash.merge";
import JSON5 from "json5";
import prettier from "prettier";
import { PgenError } from "./utils/pgen-error";
import type { Context, GeneratorOptions, InternalOptions } from "./types/index";
import { getDbObjects, resolvePath, augmentContext, composeWith, parseTemplatePath, scanTemplateDir } from "./utils";

const COLORS: Record<string, keyof typeof chalk> = {
  default: "green",
  delete: "yellow",
};

/**
 * Base abstract class for for pg-generator classes.
 */
export abstract class PgGenerator<O extends GeneratorOptions = GeneratorOptions> {
  #destinationRoot: string;
  #internalOptions: InternalOptions;
  protected fs: any;
  protected options: O & { envPrefix: string; context: Record<string, any> };

  /**
   * Creates an instance of PgGenerator.
   *
   * @param options are parameters for several options.
   * @param internalOptions are added by `pg-generator` functions and not for user.
   */
  public constructor(options: O, internalOptions: InternalOptions) {
    if (options.clear && !options.outDir) throw new PgenError("As a precaution, 'clear' option cannot be used without 'outDir' option.");
    const collisions = internalOptions.db.relationNameCollisions;
    if (collisions) throw PgenError.collisionError(collisions);

    this.options = { envPrefix: "DB", copy: true, context: {}, ...options };
    this.#internalOptions = internalOptions;
    this.#destinationRoot = resolve(internalOptions.cwd, this.options.outDir || "");
  }

  /**
   * Renders all templates in `[rootDir]/templates` directory with the render function using related context data,
   * and writes generated contents to the files in output directory. If render function returns `undefined`
   * for a template/database object pair no file is not written for that pair.
   *
   * Template file names may contain variables and basic filter functions to change names of generated files.
   *
   * Additionally copies all files in `[rootDir]/files` to the output directory.
   */
  public async generate(): Promise<void> {
    if (this.options.clear) await this.clear();
    await this.init();

    const { templates, files } = await scanTemplateDir(this.#internalOptions.templateDir);
    const context = await this.fetchContext();

    // Generate templates. Provide absolute outdir. Don't copy files, because you should use Yeoman's `mem-fs-editor` API.
    const processTemplates = Promise.all(
      templates.flatMap((path) => {
        const { className, accessor, targetPath } = parseTemplatePath(path);
        return getDbObjects(this.#internalOptions.db, className, accessor, path).map(async (dbObject) => {
          const destinationPath = resolvePath(targetPath, dbObject); // e.g. {name#dashcase}.js -> member-option.js
          const content = await this.format(destinationPath, await this.render(path, augmentContext(dbObject, context)));
          if (content !== undefined && content.replace(/\s/g, "") !== "") await this.process(destinationPath, content);
        });
      })
    );
    const copyFiles = Promise.all(files.map((file) => this.copyTemplate(file, file)));
    await Promise.all([processTemplates, copyFiles]);
  }

  //
  // ────────────────────────────────────────────────────────── I ──────────
  //   :::::: L I F E C Y C L E : :  :   :    :     :        :          :
  // ────────────────────────────────────────────────────────────────────
  //
  // Methods below are lifecycle methods and most probably would be extended
  // by child classes for custom behavior. Providing `render` is mandatory.
  // Flow: clear -> init --> context -> render -> format -> process

  /**
   * If clear option is true, this method is called. It deletes destination path.
   * User may override this method to tailor clear operation according to generator's
   * purposes.
   */
  protected async clear(): Promise<any> {
    await this.deleteDestination();
  }

  /**
   * Executed after clear operation.
   */
  protected init(): Promise<any> | any {
    return undefined;
  }

  /**
   * Provides generator specific extra context to templates. This data is deeply merged with
   * context provided by user.
   *
   * @returns generator spesific extra context data.
   */
  protected context(): Promise<Record<string, any>> | Record<string, any> {
    return {};
  }

  /**
   * Generates content from the template with provided path using the context.
   *
   * @param templatePath is the path of the template file to generate content.
   * @param context is the data to be used for generating content.
   * @returns generated content.
   */
  protected render(templatePath: string, context: Context): Promise<undefined | string> | undefined | string {
    const stub = templatePath ?? context; // eslint-disable-line @typescript-eslint/no-unused-vars
    return undefined;
  }

  /**
   * Formats given content. By default `prettier` is used. Plase note that at this stage file is not written to disk yet.
   *
   * @param destination is the path of the destination file to be created.
   * @param content is the content to format.
   * @returns formatted code.
   */
  protected format(destination: string, content?: string): Promise<undefined | string> | undefined | string {
    if (content === undefined || content === null || content === "") return content;
    return prettier.format(content, { filepath: destination });
  }

  /**
   * Writes generated content from [[render]] to the destination file.
   *
   * @param destination is the destionation path of the file to be cerated..
   * @param content is the content to process.
   */
  protected process(destination: string, content: string): Promise<any> | any {
    return this.writeTemplate(destination, content);
  }

  //
  // ────────────────────────────────────────────────────── I ──────────
  //   :::::: U T I L I T Y : :  :   :    :     :        :          :
  // ────────────────────────────────────────────────────────────────
  //

  // Protected methods below are utility methods to be used by child classes.

  /**
   * Logs given messsage if `options.log` is true.
   *
   * @param title is the title of the log.
   * @param message is the nessage to log.
   * @param options are options.
   */
  protected log(title: string, { message, source, destination }: { message?: string; source?: string; destination?: string }): void {
    if (!this.options.log) return;
    const titleColor = COLORS[title] ?? COLORS.default;
    const colorizeTitle = chalk[titleColor] as (input: string) => string;
    const logTitle = colorizeTitle(title);
    const messageSpace = message && source && destination ? ": " : " ";
    const logMessage = message === undefined ? "" : ` ${message}${messageSpace}`;
    const logSource = source ? `${source}${destination ? " ➔ " : ""}` : "";
    const logDestination = destination ? join(this.options.outDir || "", destination) : "";
    console.log(`   ${logTitle} ${logMessage}${logSource}${logDestination}`); // eslint-disable-line no-console
  }

  /**
   * Returns abosulte path for the given path relative to output directory.
   *
   * @param path is the path relative to output directory.
   * @returns the absolute path.
   */
  protected destinationPath(path = ""): string {
    return resolve(this.#destinationRoot, path);
  }

  /**
   * Returns abosulte path for the given path relative to template root.
   *
   * @param path is the path relative to template root.
   * @returns the absolute path.
   */
  protected templatePath(path = ""): string {
    return resolve(this.#internalOptions.templateDir, path);
  }

  //
  // ──────────────────────────────────────────────────────── I ──────────
  //   :::::: I N T E R N A L : :  :   :    :     :        :          :
  // ──────────────────────────────────────────────────────────────────
  //

  // Private methods below are for internal usage.

  /**
   * Executes the default sub-generator `app` from a generator. If options are provided, they are combined with current options.
   *
   * @param generator is the name or path of the generator. If it is a local path, use `require.resolve` or provide an absolute path.
   * @param options are the new options added on top of curent options.
   *
   * @throws error if generator can not be found.
   *
   * @example
   * this.composeWith("generator-from-npm");
   * this.composeWith("generator-from-npm", options);
   * this.composeWith(require.resolve("./local-generator"), options);
   */
  protected async composeWith(generator: string, options?: GeneratorOptions): Promise<void>;
  /**
   * Executes a sub-generator from a generator. If options are provided, they are merged with current options. Also some of the options
   * are changed to sensible default values for sub-generators (e.g. `clear` is changed to `false`), but they can be overridden by
   * newly provided options.
   *
   * @param generator is the name or path of the generator. If it is a local path, use `require.resolve` or provide an absolute path.
   * @param subGeneratorName is the name of the sub-generator to be executed.
   * @param options are the new options added on top of curent options.
   *
   * @throws error if generator or sub-generator can not be found.
   *
   * @example
   * this.composeWith("generator-from-npm", "sub-generator");
   * this.composeWith(require.resolve("./local-generator"), "sub-generator", options);
   */
  protected async composeWith(generator: string, subGeneratorName?: string, options?: GeneratorOptions): Promise<void>;
  protected async composeWith(generator: string, subOrOptions?: string | GeneratorOptions, maybeOptions?: GeneratorOptions): Promise<void> {
    const [options = {}, generatorName] = typeof subOrOptions === "string" ? [maybeOptions, subOrOptions] : [maybeOptions];
    const newOptions = { ...this.options, clear: false, ...options };
    return composeWith(generator, generatorName, newOptions, this.#internalOptions.db, this.#internalOptions.cwd);
  }

  /**
   * Writes given content to the file after creating it's directory.
   *
   * @param destinantion is the path to write file.
   * @param content is the content to write.
   */
  private async writeTemplate(destination: string, content: string): Promise<void> {
    const destinationPath = this.destinationPath(destination);
    await fs.mkdir(dirname(destinationPath), { recursive: true });
    await fs.writeFile(destinationPath, content, { encoding: "utf8" });
    this.log("create", { destination });
  }

  /**
   * Copy the source file to the destination. Also creates the destination directory if it does not exist.
   *
   * @param source is the source file to copy.
   * @param destination is the destination file to be created.
   */
  private async copyTemplate(source: string, destination: string): Promise<void> {
    await fs.mkdir(dirname(this.destinationPath(destination)), { recursive: true });
    await fs.copyFile(this.templatePath(source), this.destinationPath(destination));
    this.log("create", { destination });
  }

  /**
   * Removes file or directory (and it's contents() from destination. If destination is equal to `cwd()`
   * it throws error and aborts operation for a safety measure. Most probably `cwd()` contains other files too.
   *
   * @param path is the path relative to the destination root.
   */
  private async deleteDestination(path = ""): Promise<void> {
    const destination = this.destinationPath(path);
    const isRoot = resolve(this.#internalOptions.cwd) === resolve(destination);
    if (isRoot) throw new PgenError("Root is not deleted, because it may be a mistake. Probably it would contain other files.");

    try {
      await fs.rmdir(this.destinationPath(path), { recursive: true });
    } catch (error) {
      if (error.code !== "ENOENT") throw error;
      else if (error.code === "ENOTDIR") await fs.unlink(this.destinationPath(path));
    }
    this.log("delete", { destination: path || "." });
  }

  /**
   * Fetches and merges context data from `pg-generator` instance, context file and options.
   * The context file may be a JSON5 data file (may contain comments etc.) or a JavaScript file
   * with a default export of object or a function. If it is a function, it is passed
   * `pg-structure` db instance.
   *
   * Merge priorities are: `options.context` > context file > `context()` method.
   *
   * @param contextFile is the file to read from.
   * @param context is the context data.
   * @returns merged context data.
   */
  private async fetchContext(): Promise<Record<string, any>> {
    let contextFromFile = {};

    if (this.options.contextFile !== undefined) {
      const contextFile = resolve(this.options.contextFile);
      const fileContent =
        extname(contextFile) === ".json"
          ? JSON5.parse(await fs.readFile(contextFile, { encoding: "utf8" }))
          : (await import(contextFile)).default;
      contextFromFile = typeof fileContent === "function" ? fileContent(this.#internalOptions.db) : fileContent;

      this.log("context", { destination: contextFile });
    }

    return merge(await this.context(), contextFromFile, this.options.context);
  }
}
