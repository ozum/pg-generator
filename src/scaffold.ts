import { join, basename, extname } from "path";
import { EOL } from "os";
import { promises as fs } from "fs";
import { ignoreCode } from "ignor";
import { PgenError } from "./utils/pgen-error";
import { copyDir } from "./utils";

const SCAFFOLD_DIR = join(__dirname, "../module-files/scaffolds");

/**
 * Renames "name" entry of the "package.json" in given directory.
 *
 * @param dir is the directory that contains the "package.json" file.
 * @param newName is the new name.
 */
async function renameInPackageJson(dir: string, newName: string): Promise<void> {
  const pkgPath = join(dir, "package.json");
  const pkg = JSON.parse(await fs.readFile(pkgPath, { encoding: "utf-8" }));
  pkg.name = newName;
  await fs.writeFile(pkgPath, JSON.stringify(pkg, undefined, 2));
}

/**
 * Formats given list of scaffold names.
 *
 * @param scaffolds are the list of the scaffold names.
 * @internal
 */
export function getScaffoldsMessage(scaffolds: string[]): string {
  const available = scaffolds.join(`${EOL}  ● `);
  return `Available scaffolds are:${EOL}  ● ${available}`;
}

/**
 * Read list of scaffolds from disk.
 *
 * @internal
 */
export async function readScaffolds(): Promise<string[]> {
  return (await fs.readdir(SCAFFOLD_DIR, { withFileTypes: true })).filter((e) => e.isDirectory()).map((e) => e.name);
}

/**
 * Creates a skeleton Node.js project for a generator (pg-generator plugin).
 *
 * @param scaffoldName is the name to create scaffold from.
 * @param options are options
 * @param options.outDir is the output directory to create the project.
 * @param options.name is the name of the created project.
 * @throws error if required scaffold is unavailable or output directory contains files.
 */
export async function scaffold(
  scaffoldName: string,
  { outDir, name = `pg-generator-${basename(outDir, extname(outDir)).replace("pg-generator-", "")}` }: { outDir: string; name?: string }
): Promise<void> {
  const scaffolds = await readScaffolds();
  if (scaffoldName === undefined) throw new PgenError(`SCAFFOLD is required. ${getScaffoldsMessage(scaffolds)}`);
  if (!scaffolds.includes(scaffoldName)) throw new PgenError(`Unknown scaffold "${scaffold}". ${getScaffoldsMessage(scaffolds)}`);

  try {
    const destinationEntries = await fs.readdir(outDir).catch(ignoreCode("ENOENT"));
    if (destinationEntries && destinationEntries.length > 0) throw new PgenError(`Output directory "${outDir}" is not empty.`);
  } catch (error) {
    if (error.code === "ENOTDIR") throw new PgenError(`"${outDir}" is not a directory.`);
    throw error;
  }

  await copyDir(join(SCAFFOLD_DIR, scaffoldName), outDir);
  await renameInPackageJson(outDir, name);
}
