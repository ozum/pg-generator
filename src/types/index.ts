import type { DbObject, Db, Options as PgStructureOptions } from "pg-structure";
import type { Client, ClientConfig } from "pg"; // eslint-disable-line import/no-unresolved

/** Options for generation and reverse engineering process. Options extends [pg-structure options](https://www.pg-structure.com/nav.02.api/interfaces/options) */
export interface GeneratorOptions {
  /** Whether to clear the destination directory before creating files. */
  clear?: boolean;
  /** Path of the output directory. */
  outDir?: string;
  /** Path to a JSON or JS file providing extra context for templates. */
  contextFile?: string;
  /** Extra context data. This data is merged with and overridden by data from context file. */
  context?: Record<string, any>;
  /** Whether to log output to console. */
  log?: boolean;
}

export interface ClientOptions extends PgStructureOptions {
  /** Either a [node-postgres client](https://node-postgres.com/api/client) or a configuration object or a connection string to create a [node-postgres client](https://node-postgres.com/api/client). */
  client?: Client | ClientConfig | string;

  // ─── PG CLIENT OPTIONS ──────────────────────────────────────────────────────────

  /** Database username. Default from environment var: PGUSER || USER || DB_USER */
  user?: string;
  /** Database password. Default from environment var: PGPASSWORD || DB_PASSWORD */
  password?: string;
  /** Database host. Default from environment var: PGHOST || DB_HOST */
  host?: string;
  /** Database to connect. Default from environment var: PGDATABASE || DB_DATABASE */
  database?: string;
  /** Database port. Default from environment var: PGPORT || DB_PORT */
  port?: string;
  /** Connection string to connect to the database e.g. postgres://user:password@host:5432/database */
  connectionString?: string;
  /** Passed directly to node.TLSSocket, supports all tls.connect options */
  ssl?: any;
}

/** Extended options which includes internal options added by `pg-generator` not by user. */
export interface InternalOptions {
  /** The path of the directory that contains template files. */
  templateDir: string;
  /** The path of the cwd . If outpur directory is a relative path, it is calculated relative to destination root. */
  cwd: string;
  /** [pg-structure Db instance](https://www.pg-structure.com/nav.02.api/classes/db)  */
  db: Db;
}

export type Options = GeneratorOptions & ClientOptions;

/** Context provided to render function. */
export type Context = { o: Db | DbObject; x: Record<string, any> };
