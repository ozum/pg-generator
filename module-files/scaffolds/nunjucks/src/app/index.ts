/* eslint-disable class-methods-use-this */
import { PgGenerator, Context } from "pg-generator";
import { getNunjucks } from "../utils/get-nunjucks";
import generatorContext from "./context";

const nunjucks = getNunjucks(__dirname);

/**
 * Example `pg-generator` generator which creates sequelize models.
 */
export default class App extends PgGenerator {
  /**
   * Returns a context data including sequelize data types and some other examples.
   */
  protected context(): Record<string, any> {
    return generatorContext;
  }

  /**
   * Render method which uses a nunjucks template. You can use any template engine.
   * `pg-generator` is template engine agnostic.
   */
  protected async render(templatePath: string, context: Context): Promise<string> {
    return nunjucks.render(templatePath, context);
  }
}
