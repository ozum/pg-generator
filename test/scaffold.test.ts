import { join } from "path";
import { promises as fs } from "fs";
import { tmpdir } from "os";
import { load } from "fs-structure";
import { scaffold } from "../src/index";

const TEST_ROOT = join(tmpdir(), "pgen");
const outDir = join(TEST_ROOT, "out");

afterEach(async () => {
  await fs.rmdir(outDir, { recursive: true });
});

afterAll(async () => {
  await fs.rmdir(TEST_ROOT, { recursive: true });
});

describe("scaffold", () => {
  it("should generate nunjucks scaffold.", async () => {
    await scaffold("nunjucks", { outDir });
    const structure = await load(outDir);
    expect(structure).toMatchSnapshot();
  });

  it("should throw if no scaffold is provided.", async () => {
    await expect(() => scaffold(undefined as any, { outDir })).rejects.toThrow("is required");
  });

  it("should throw if unknown scaffold is provided.", async () => {
    await expect(() => scaffold("NON EXISTING SCAFFOLD", { outDir })).rejects.toThrow("Unknown scaffold");
  });

  it("should throw if destination directory has files in it.", async () => {
    await expect(() => scaffold("nunjucks", { outDir: join(__dirname, "test-helper/non-empty-dir") })).rejects.toThrow("is not empty");
  });

  it("should throw if destination directory is not a directory.", async () => {
    await expect(() => scaffold("nunjucks", { outDir: join(__dirname, "test-helper/non-empty-dir/file.txt") })).rejects.toThrow(
      "is not a directory"
    );
  });
});
