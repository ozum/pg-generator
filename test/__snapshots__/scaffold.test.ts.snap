// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scaffold should generate nunjucks scaffold. 1`] = `
Object {
  "package.json": "{
  \\"name\\": \\"pg-generator-out\\",
  \\"version\\": \\"1.0.0\\",
  \\"description\\": \\"pg-generator plugin that generates ...\\",
  \\"author\\": \\"\\",
  \\"keywords\\": [
    \\"pg-generator\\",
    \\"plugin\\"
  ],
  \\"main\\": \\"dist/index.js\\",
  \\"files\\": [
    \\"dist\\"
  ],
  \\"types\\": \\"dist/index.d.ts\\",
  \\"scripts\\": {
    \\"build\\": \\"tsc\\",
    \\"test\\": \\"echo \\\\\\"Error: no test specified\\\\\\" && exit 1\\"
  },
  \\"homepage\\": \\"\\",
  \\"bugs\\": \\"\\",
  \\"repository\\": \\"\\",
  \\"license\\": \\"MIT\\",
  \\"dependencies\\": {
    \\"nunjucks\\": \\"^3.2.3\\",
    \\"pg-generator\\": \\"^5.0.0-alpha.6\\"
  },
  \\"devDependencies\\": {
    \\"@types/node\\": \\"^14.14.31\\",
    \\"@types/nunjucks\\": \\"^3.1.4\\",
    \\"pg-structure\\": \\"^7.7.1\\",
    \\"typescript\\": \\"^4.1.5\\"
  }
}",
  "src/app/context.ts": "export default {
  // Values under \\"global\\" key is sent to all templates.
  global: {
    // We added an attribute called \\"addSchemaName\\". It has no special meaning, and you can choose any name.
    // We used \\"addSchemaName\\" in templates to determine whether to add schema names to class names.
    addSchemaName: true,
  },
  // Values under \\"tables\\" key is sent to templates related to corresponding table.
  tables: {
    // Data below only sent to templates related to \`public.xyz\` table.*/
    \\"public.xyz\\": {},
  },
};
",
  "src/app/index.ts": "/* eslint-disable class-methods-use-this */
import { PgGenerator, Context } from \\"pg-generator\\";
import { getNunjucks } from \\"../utils/get-nunjucks\\";
import generatorContext from \\"./context\\";

const nunjucks = getNunjucks(__dirname);

/**
 * Example \`pg-generator\` generator which creates sequelize models.
 */
export default class App extends PgGenerator {
  /**
   * Returns a context data including sequelize data types and some other examples.
   */
  protected context(): Record<string, any> {
    return generatorContext;
  }

  /**
   * Render method which uses a nunjucks template. You can use any template engine.
   * \`pg-generator\` is template engine agnostic.
   */
  protected async render(templatePath: string, context: Context): Promise<string> {
    return nunjucks.render(templatePath, context);
  }
}
",
  "src/app/templates/[db] index.js.njk": "{%- for table in db.tables %}
const {{ table | dboClassName(addSchemaName) }} = \\"\\";
{%- endfor %}

",
  "src/app/templates/[table] {schema.name # dash-case}.{name # dash-case}.js.njk": "{%- for column in table.columns %}
{{ column.name }}
{%- endfor %}

",
  "src/index.ts": "console.warn(\\"This is a pg-generator plugin. It is not intended for direct usage. Use it vie pg-generator.\\"); // eslint-disable-line no-console
",
  "src/partials/stub.md.njk": "This is a stub file to commit the directory to git.
",
  "src/utils/filters.ts": "import type { Table } from \\"pg-structure\\";

/**
 * Adds \\"Hello!\\"\\" to the beginning of the string.
 *
 * @param input is the input.
 * @returns the text with \\"Hello!\\".
 *
 * @example
 * {{ table.name | hello }}
 */
export function hello(input: string): string {
  return \`Hello! \${input}\`;
}

/**
 * Returns column names of the table as a CSV list.
 *
 * @param table is the table to get columns of.
 * @returns column names as CSV.
 *
 * @example
 * {{ table | columnNames }}
 */
export function columnNames(table: Table): string {
  return table.columns.map((column) => column.name).join(\\", \\");
}
",
  "src/utils/get-nunjucks.ts": "import { join } from \\"path\\";
import { filterFunctions } from \\"pg-generator\\";
import nunjucks, { Environment } from \\"nunjucks\\";
import * as filters from \\"./filters\\";

/**
 * Creates \`nunjucks\` template engine environment for the given generator and adds filters.
 *
 * @param generatorPath is the path of the generator.
 */
export function getNunjucks(generatorPath: string): Environment {
  // Create \`nunjucks\` environment.
  const environment = new nunjucks.Environment(
    // Add given sub-generator path and shared \\"partials\\" path to the list of template paths.
    new nunjucks.FileSystemLoader([join(generatorPath, \\"templates\\"), join(__dirname, \\"../partials\\")]),
    { autoescape: false }
  );

  // Add pg-generator and custom filters to nunjucks. Filters are not nunjucks specific, many template engines have a similar mechanism.
  Object.entries({ ...filterFunctions, ...filters }).forEach(([name, filter]) => environment.addFilter(name, filter));
  return environment;
}
",
  "tsconfig.json": "{
  \\"compilerOptions\\": {
    // Visit https://aka.ms/tsconfig.json to read more about this file

    // Basic Options
    \\"incremental\\": true, // Enable incremental compilation
    \\"target\\": \\"ES2019\\", // Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'.
    \\"module\\": \\"commonjs\\", // Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'.
    // \\"lib\\": [\\"dom\\", \\"esnext\\"],              // Specify library files to be included in the compilation.
    // \\"allowJs\\": true,                       // Allow javascript files to be compiled.
    // \\"checkJs\\": true,                       // Report errors in .js files.
    // \\"jsx\\": \\"preserve\\",                     // Specify JSX code generation: 'preserve', 'react-native', or 'react'.
    \\"declaration\\": true, // Generates corresponding '.d.ts' file.
    \\"declarationMap\\": true, // Generates a sourcemap for each corresponding '.d.ts' file.

    \\"sourceMap\\": true, // Generates corresponding '.map' file.
    // \\"outFile\\": \\"./\\",                       // Concatenate and emit output to single file.
    \\"outDir\\": \\"dist\\", // Redirect output structure to the directory.
    // \\"rootDir\\": \\"./\\",                       // Specify the root directory of input files. Use to control the output directory structure with --outDir.
    // \\"composite\\": true,                     // Enable project compilation
    // \\"tsBuildInfoFile\\": \\"./\\",               // Specify file to store incremental compilation information
    // \\"removeComments\\": true,                // Do not emit comments to output.
    // \\"noEmit\\": true,                        // Do not emit outputs.
    // \\"importHelpers\\": true,                 // Import emit helpers from 'tslib'.
    // \\"downlevelIteration\\": true,            // Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'.
    // \\"isolatedModules\\": true,               // Transpile each file as a separate module (similar to 'ts.transpileModule').

    // Strict Type-Checking Options
    \\"strict\\": true, // Enable all strict type-checking options.
    // \\"noImplicitAny\\": true,                 // Raise error on expressions and declarations with an implied 'any' type.
    // \\"strictNullChecks\\": true,              // Enable strict null checks.
    // \\"strictFunctionTypes\\": true,           // Enable strict checking of function types.
    // \\"strictBindCallApply\\": true,           // Enable strict 'bind', 'call', and 'apply' methods on functions.
    // \\"strictPropertyInitialization\\": true,  // Enable strict checking of property initialization in classes.
    // \\"noImplicitThis\\": true,                // Raise error on 'this' expressions with an implied 'any' type.
    // \\"alwaysStrict\\": true,                  // Parse in strict mode and emit \\"use strict\\" for each source file.

    // Additional Checks
    // \\"noUnusedLocals\\": true,                // Report errors on unused locals.
    // \\"noUnusedParameters\\": true,            // Report errors on unused parameters.
    // \\"noImplicitReturns\\": true,             // Report error when not all code paths in function return a value.
    // \\"noFallthroughCasesInSwitch\\": true,    // Report errors for fallthrough cases in switch statement.
    // \\"noUncheckedIndexedAccess\\": true,      // Include 'undefined' in index signature results

    // Module Resolution Options
    // \\"moduleResolution\\": \\"node\\",            // Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6).
    // \\"baseUrl\\": \\"./\\",                       // Base directory to resolve non-absolute module names.
    // \\"paths\\": {},                           // A series of entries which re-map imports to lookup locations relative to the 'baseUrl'.
    // \\"rootDirs\\": [],                        // List of root folders whose combined content represents the structure of the project at runtime.
    // \\"typeRoots\\": [],                       // List of folders to include type definitions from.
    // \\"types\\": [],                           // Type declaration files to be included in compilation.
    // \\"allowSyntheticDefaultImports\\": true,  // Allow default imports from modules with no default export. This does not affect code emit, just typechecking.
    \\"esModuleInterop\\": true, // Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'.
    // \\"preserveSymlinks\\": true,              // Do not resolve the real path of symlinks.
    // \\"allowUmdGlobalAccess\\": true,          // Allow accessing UMD globals from modules.

    // Source Map Options
    // \\"sourceRoot\\": \\"\\",                      // Specify the location where debugger should locate TypeScript files instead of source locations.
    // \\"mapRoot\\": \\"\\",                         // Specify the location where debugger should locate map files instead of generated locations.
    // \\"inlineSourceMap\\": true,               // Emit a single file with source maps instead of having a separate file.
    // \\"inlineSources\\": true,                 // Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set.

    // Experimental Options
    \\"experimentalDecorators\\": true, // Enables experimental support for ES7 decorators.
    \\"emitDecoratorMetadata\\": true, // Enables experimental support for emitting type metadata for decorators.

    // Advanced Options
    \\"skipLibCheck\\": true, // Skip type checking of declaration files.
    \\"forceConsistentCasingInFileNames\\": true // Disallow inconsistently-cased references to the same file.
  },
  \\"include\\": [\\"src/**/*\\"],
  \\"exclude\\": [\\"node_modules\\", \\"dist\\"]
}
",
}
`;
